src/computedevs.jl

```diff
@@ -0,0 +1,80 @@
+# adaptors for converting abstract arrays to Base.Array
+
+
+"""
+    abstract type AbstractComputingDevice
+
+Supertype for CPU and GPU computing devices.
+
+`adapt(dev::AbstractComputingDevice, x)` adapts `x` for `dev`.
+"""
+abstract type AbstractComputingDevice end
+export AbstractComputingDevice
+
+
+"""
+    struct ComputingDeviceIndependent <: AbstractComputingDevice
+
+`get_computing_device(x) === ComputingDeviceIndependent()` indicates
+that `x` is not tied to a specific computing device. This typically
+means that x is a statically allocated object.
+"""
+struct ComputingDeviceIndependent <: AbstractComputingDevice end
+export ComputingDeviceIndependent
+
+
+"""
+    struct CPUDevice <: AbstractComputingDevice
+
+`CPUDevice()` is the default CPU device.
+"""
+struct CPUDevice <: AbstractComputingDevice end
+export CPUDevice
```

Comment:
Do we need all of these as exports? "CPUDevice" is pretty non-specific as a name, and could clash with some other package that is aware of computing devices (nothing comes to mind, though).

Reply:
Oceanigans has his own hierarchy as an example. @glwagner

Reply:
> Do we need all of these as exports? "CPUDevice" is pretty non-specific as a name, and could clash with some other package that is aware of computing devices (nothing comes to mind, though).

I didn't find any package that uses `CPUDevice` currently. I'm always for avoiding name clashes - but since the name is "free": The advantage of exporting this would be to encourage packages to use this new common infrastructure instead of rolling their own, so to speak. :-)

Reply:
I'll mark this resolved since a lot has changed in the PR, please feel free to reopen.